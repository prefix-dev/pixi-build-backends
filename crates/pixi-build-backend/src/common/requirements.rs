use std::collections::{BTreeMap, HashMap};

use rattler_build::{
    NormalizedKey,
    recipe::{parser::Requirements, variable::Variable},
};
use serde::Serialize;

use crate::{
    PackageSpec, ProjectModel, Targets, dependencies::ExtractedDependencies, traits::Dependencies,
};

pub struct PackageRequirements<P: ProjectModel> {
    /// Requirements for rattler-build
    pub requirements: Requirements,

    /// The source requirements
    pub source: SourceRequirements<P>,
}

#[derive(Debug, Serialize)]
#[serde(bound(
    serialize = "<<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec: Serialize"
))]
pub struct SourceRequirements<P: ProjectModel> {
    /// Source package specification for build dependencies
    pub build: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,

    /// Source package specification for host dependencies
    pub host: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,

    /// Source package specification for runtime dependencies
    pub run: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,
}

/// Return requirements for the given project model
pub fn requirements<P: ProjectModel>(
    dependencies: Dependencies<<P::Targets as Targets>::Spec>,
    variant: &BTreeMap<NormalizedKey, Variable>,
) -> miette::Result<PackageRequirements<P>> {
    let build = ExtractedDependencies::from_dependencies(dependencies.build, variant)?;
    let host = ExtractedDependencies::from_dependencies(dependencies.host, variant)?;
    let run = ExtractedDependencies::from_dependencies(dependencies.run, variant)?;

    Ok(PackageRequirements {
        requirements: Requirements {
            build: build.dependencies,
            host: host.dependencies,
            run: run.dependencies,
            ..Default::default()
        },
        source: SourceRequirements {
            build: build.sources,
            host: host.sources,
            run: run.sources,
        },
    })
}
