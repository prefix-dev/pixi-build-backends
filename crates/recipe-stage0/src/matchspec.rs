use rattler_conda_types::{MatchSpec, PackageName, ParseMatchSpecError};
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::{fmt::Display, str::FromStr};
use url::Url;

// Wrapper for MatchSpec to enable serde support
#[derive(Debug, Clone, Default)]
pub struct SerializableMatchSpec(pub MatchSpec);

impl Serialize for SerializableMatchSpec {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.0.to_string())
    }
}

impl<'de> Deserialize<'de> for SerializableMatchSpec {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        MatchSpec::from_str(&s, rattler_conda_types::ParseStrictness::Strict)
            .map(SerializableMatchSpec)
            .map_err(serde::de::Error::custom)
    }
}

impl From<MatchSpec> for SerializableMatchSpec {
    fn from(spec: MatchSpec) -> Self {
        SerializableMatchSpec(spec)
    }
}

impl From<&str> for SerializableMatchSpec {
    fn from(s: &str) -> Self {
        SerializableMatchSpec(
            MatchSpec::from_str(s, rattler_conda_types::ParseStrictness::Strict)
                .expect("Invalid MatchSpec"),
        )
    }
}

impl From<String> for SerializableMatchSpec {
    fn from(s: String) -> Self {
        SerializableMatchSpec(
            MatchSpec::from_str(&s, rattler_conda_types::ParseStrictness::Strict)
                .expect("Invalid MatchSpec"),
        )
    }
}

impl Display for SerializableMatchSpec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FromStr for SerializableMatchSpec {
    type Err = ParseMatchSpecError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        MatchSpec::from_str(s, rattler_conda_types::ParseStrictness::Strict)
            .map(SerializableMatchSpec)
    }
}

#[derive(Clone, PartialEq, Debug)]
pub struct SourceMatchSpec {
    pub spec: MatchSpec,
    pub location: Url,
}

#[derive(Clone, PartialEq, Debug)]
pub enum PackageDependency {
    Binary(MatchSpec),
    Source(SourceMatchSpec),
}

impl PackageDependency {
    pub fn package_name(&self) -> PackageName {
        match self {
            PackageDependency::Binary(spec) => {
                spec.name.clone().expect("Binary spec should have a name")
            }
            PackageDependency::Source(source_spec) => source_spec
                .spec
                .name
                .clone()
                .expect("Source spec should have a name"),
        }
    }

    pub fn as_source(&self) -> Option<&SourceMatchSpec> {
        if let PackageDependency::Source(source_spec) = self {
            Some(source_spec)
        } else {
            None
        }
    }

    /// Check if the dependency can be used as a variant in a recipe.
    pub fn can_be_used_as_variant(&self) -> bool {
        match self {
            PackageDependency::Binary(boxed_spec) => {
                let rattler_conda_types::MatchSpec {
                    version,
                    build,
                    build_number,
                    file_name,
                    channel,
                    subdir,
                    md5,
                    sha256,
                    ..
                } = boxed_spec;

                version == &Some(rattler_conda_types::VersionSpec::Any)
                    && build.is_none()
                    && build_number.is_none()
                    && file_name.is_none()
                    && channel.is_none()
                    && subdir.is_none()
                    && md5.is_none()
                    && sha256.is_none()
            }
            _ => false,
        }
    }
}

impl Display for PackageDependency {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PackageDependency::Binary(spec) => write!(f, "{}", spec),
            PackageDependency::Source(source_spec) => {
                let matchspec = SerializableMatchSpec::from(source_spec.spec.clone());
                write!(f, "{}", matchspec)
            }
        }
    }
}

impl From<SerializableMatchSpec> for PackageDependency {
    fn from(spec: SerializableMatchSpec) -> Self {
        // we need to split on url to determine if this is a binary or source dependency
        if let Some(url) = spec.0.url.as_ref() {
            // remove the URL from the MatchSpec
            let mut spec = spec.0.clone();
            spec.url = None;

            PackageDependency::Source(SourceMatchSpec {
                spec,
                location: url.clone(),
            })
        } else {
            PackageDependency::Binary(spec.0)
        }
    }
}

impl FromStr for PackageDependency {
    type Err = ParseMatchSpecError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        SerializableMatchSpec::from_str(s).map(PackageDependency::from)
    }
}

impl From<&str> for PackageDependency {
    fn from(s: &str) -> Self {
        SerializableMatchSpec::from(s).into()
    }
}

impl From<PackageDependency> for SerializableMatchSpec {
    fn from(val: PackageDependency) -> Self {
        match val {
            PackageDependency::Binary(spec) => SerializableMatchSpec(spec),
            PackageDependency::Source(source_spec) => {
                // we need to put the URL from source spec back into the MatchSpec

                let mut matchspec = source_spec.spec.clone();
                matchspec.url = Some(source_spec.location);

                SerializableMatchSpec(matchspec)
            }
        }
    }
}

impl Serialize for PackageDependency {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let dep = self.clone();

        let ser_matchspec: SerializableMatchSpec = dep.into();
        serializer.serialize_str(&ser_matchspec.to_string())
    }
}

impl<'de> Deserialize<'de> for PackageDependency {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        MatchSpec::from_str(&s, rattler_conda_types::ParseStrictness::Strict)
            .map(SerializableMatchSpec)
            .map(PackageDependency::from)
            .map_err(serde::de::Error::custom)
    }
}
