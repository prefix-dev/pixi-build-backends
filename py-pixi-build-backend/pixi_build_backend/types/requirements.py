from typing import Optional
from pixi_build_backend.pixi_build_backend import PyPackageDependency, PySourceMatchSpec


class SourceMatchSpec:
    """A source match spec wrapper."""

    _inner: PySourceMatchSpec

    def __init__(self, spec: str, location: str) -> None:
        self._inner = PySourceMatchSpec(spec, location)

    @property
    def spec(self) -> str:
        """Get the spec."""
        return self._inner.spec

    @property
    def location(self) -> str:
        """Get the location."""
        return self._inner.location

    def _into_py(self) -> PySourceMatchSpec:
        """Convert to PySourceMatchSpec."""
        return self._inner

    @classmethod
    def _from_inner(cls, inner: PySourceMatchSpec) -> "SourceMatchSpec":
        """Create from PySourceMatchSpec."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class PackageDependency:
    """A package dependency wrapper."""

    _inner: PyPackageDependency

    def __init__(self, spec: str) -> None:
        self._inner = PyPackageDependency(spec)


    @property
    def is_binary(self) -> bool:
        """Check if this is a binary dependency."""
        return self._inner.is_binary()

    @property
    def is_source(self) -> bool:
        """Check if this is a source dependency."""
        return self._inner.is_source()

    @property
    def binary_spec(self) -> Optional[str]:
        """Get the binary spec if this is a binary dependency."""
        return self._inner.get_binary()

    @property
    def source_spec(self) -> Optional[SourceMatchSpec]:
        """Get the source spec if this is a source dependency."""
        inner_spec = self._inner.get_source()
        return SourceMatchSpec._from_inner(inner_spec) if inner_spec else None

    @property
    def package_name(self) -> str:
        """Get the package name."""
        return self._inner.package_name()

    def _into_py(self) -> PyPackageDependency:
        """Convert to PyPackageDependency."""
        return self._inner

    @classmethod
    def _from_inner(cls, inner: PyPackageDependency) -> "PackageDependency":
        """Create from PyPackageDependency."""
        return cls(inner)