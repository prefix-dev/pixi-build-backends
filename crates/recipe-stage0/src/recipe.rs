use indexmap::IndexMap;
use rattler_conda_types::package::EntryPoint;
use rattler_conda_types::{PackageName, Platform};
use serde::{Deserialize, Serialize};
use std::fmt::{Debug, Display};
use std::str::FromStr;

use crate::matchspec::{PackageDependency, SerializableMatchSpec};
use crate::requirements::PackageSpecDependencies;

// Core enum for values that can be either concrete or templated
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Value<T> {
    Concrete(T),
    Template(String), // Jinja template like "${{ name|lower }}"
}

impl<T: ToString> Display for Value<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Value::Concrete(val) => write!(f, "{}", val.to_string()),
            Value::Template(template) => write!(f, "{}", template),
        }
    }
}

impl<T: ToString> Value<T> {
    pub fn concrete(&self) -> Option<&T> {
        if let Value::Concrete(val) = self {
            Some(val)
        } else {
            None
        }
    }
}

impl<T: ToString + FromStr> FromStr for Value<T>
where
    T::Err: std::fmt::Display,
{
    type Err = T::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.contains("${{") {
            // If it contains some template syntax, treat it as a template
            return Ok(Value::Template(s.to_string()));
        }

        Ok(Value::Concrete(T::from_str(s)?))
    }
}

impl From<SerializableMatchSpec> for Value<SerializableMatchSpec> {
    fn from(spec: SerializableMatchSpec) -> Self {
        Value::Concrete(spec)
    }
}

// Any item in a list can be either a value or a conditional
#[derive(Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Item<T> {
    Value(Value<T>),
    Conditional(Conditional<T>),
}

impl<T: PartialEq> PartialEq for Item<T> {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Item::Value(Value::Concrete(a)), Item::Value(Value::Concrete(b))) => a == b,
            (Item::Conditional(a), Item::Conditional(b)) => {
                a.condition == b.condition && a.then == b.then && a.else_value == b.else_value
            }
            _ => false,
        }
    }
}

impl<T: Debug> Debug for Item<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Item::Value(value) => write!(f, "Value({:?})", value),
            Item::Conditional(cond) => write!(f, "Conditional({:?})", cond),
        }
    }
}

impl<T> From<Conditional<T>> for Item<T> {
    fn from(value: Conditional<T>) -> Self {
        Self::Conditional(value)
    }
}

impl From<Source> for Item<Source> {
    fn from(source: Source) -> Self {
        Item::Value(Value::Concrete(source))
    }
}

impl From<SerializableMatchSpec> for Item<SerializableMatchSpec> {
    fn from(matchspec: SerializableMatchSpec) -> Self {
        Item::Value(Value::Concrete(matchspec))
    }
}

impl From<PackageDependency> for Item<PackageDependency> {
    fn from(dep: PackageDependency) -> Self {
        Item::Value(Value::Concrete(dep))
    }
}

impl<T: ToString + FromStr> FromStr for Item<T>
where
    T::Err: std::fmt::Display,
{
    type Err = T::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.contains("${{") {
            // If it contains some template syntax, treat it as a template
            return Ok(Item::Value(Value::Template(s.to_string())));
        }

        let value = Value::Concrete(T::from_str(s)?);
        Ok(Item::Value(value))
    }
}
#[derive(Clone)]
pub struct ListOrItem<T>(pub Vec<T>);

impl<T> Default for ListOrItem<T> {
    fn default() -> Self {
        ListOrItem(Vec::new())
    }
}

impl<T: PartialEq> PartialEq for ListOrItem<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl<T: Debug> Debug for ListOrItem<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.0.is_empty() {
            write!(f, "ListOrItem([])")
        } else if self.0.len() == 1 {
            write!(f, "ListOrItem({:?})", self.0[0])
        } else {
            write!(f, "ListOrItem({:?})", self.0)
        }
    }
}

impl<T: FromStr> FromStr for ListOrItem<T> {
    type Err = T::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(ListOrItem::single(s.parse()?))
    }
}

impl<T> serde::Serialize for ListOrItem<T>
where
    T: serde::Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self.0.len() {
            1 => self.0[0].serialize(serializer),
            _ => self.0.serialize(serializer),
        }
    }
}

impl<'de, T: serde::Deserialize<'de>> serde::Deserialize<'de> for ListOrItem<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        use serde::de::{Error, Visitor};
        use std::fmt;

        struct ListOrItemVisitor<T>(std::marker::PhantomData<T>);

        impl<'de, T: serde::Deserialize<'de>> Visitor<'de> for ListOrItemVisitor<T> {
            type Value = ListOrItem<T>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a single item or a list of items")
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::SeqAccess<'de>,
            {
                let mut vec = Vec::new();
                while let Some(item) = seq.next_element()? {
                    vec.push(item);
                }
                Ok(ListOrItem(vec))
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: Error,
            {
                let item = T::deserialize(serde::de::value::StrDeserializer::new(value))?;
                Ok(ListOrItem(vec![item]))
            }

            fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
            where
                E: Error,
            {
                let item = T::deserialize(serde::de::value::StringDeserializer::new(value))?;
                Ok(ListOrItem(vec![item]))
            }

            fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::MapAccess<'de>,
            {
                let item = T::deserialize(serde::de::value::MapAccessDeserializer::new(map))?;
                Ok(ListOrItem(vec![item]))
            }
        }

        deserializer.deserialize_any(ListOrItemVisitor(std::marker::PhantomData))
    }
}

impl<T: ToString> Display for ListOrItem<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0.len() {
            0 => write!(f, "[]"),
            1 => write!(f, "{}", self.0[0].to_string()),
            _ => write!(
                f,
                "[{}]",
                self.0
                    .iter()
                    .map(|x| x.to_string())
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        }
    }
}

impl<T> ListOrItem<T> {
    pub fn new(items: Vec<T>) -> Self {
        Self(items)
    }

    pub fn single(item: T) -> Self {
        Self(vec![item])
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn len(&self) -> usize {
        self.0.len()
    }

    pub fn iter(&self) -> std::slice::Iter<T> {
        self.0.iter()
    }
}

// Conditional structure for if-else logic
#[derive(Clone, Serialize, Deserialize)]
pub struct Conditional<T> {
    #[serde(rename = "if")]
    pub condition: String,
    pub then: ListOrItem<T>,
    #[serde(rename = "else")]
    pub else_value: ListOrItem<T>,
}

impl<T: Debug> Debug for Conditional<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Conditional {{ condition: {}, then: {:?}, else: {:?} }}",
            self.condition, self.then, self.else_value
        )
    }
}

/// Type alias for lists that can contain conditionals
pub type ConditionalList<T> = Vec<Item<T>>;

// Main recipe structure
#[derive(Serialize, Deserialize, Default)]
pub struct IntermediateRecipe {
    pub context: IndexMap<String, Value<String>>,
    pub package: Package,
    pub source: ConditionalList<Source>,
    pub build: Build,
    pub requirements: ConditionalRequirements,
    pub tests: Vec<Test>,
    pub about: Option<About>,
    pub extra: Option<Extra>,
}

pub struct EvaluatedDependencies {
    pub build: Option<Vec<SerializableMatchSpec>>,
    pub host: Option<Vec<SerializableMatchSpec>>,
    pub run: Option<Vec<SerializableMatchSpec>>,
    pub run_constraints: Option<Vec<SerializableMatchSpec>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Package {
    pub name: Value<String>,
    pub version: Value<String>,
}

impl Default for Package {
    fn default() -> Self {
        Package {
            name: Value::Concrete("default-package".to_string()),
            version: Value::Concrete("0.0.1".to_string()),
        }
    }
}

/// Source information.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Source {
    /// Url source pointing to a tarball or similar to retrieve the source from
    Url(UrlSource),
    /// Path source pointing to a local path where the source can be found
    Path(PathSource),
}

impl Source {
    pub fn url(url: String) -> Self {
        Source::Url(UrlSource {
            url: Value::Concrete(url),
            sha256: None,
        })
    }

    pub fn path(path: String) -> Self {
        Source::Path(PathSource {
            path: Value::Concrete(path),
            sha256: None,
        })
    }

    pub fn with_sha256(self, sha256: String) -> Self {
        match self {
            Source::Url(mut url_source) => {
                url_source.sha256 = Some(Value::Concrete(sha256));
                Source::Url(url_source)
            }
            Source::Path(mut path_source) => {
                path_source.sha256 = Some(Value::Concrete(sha256));
                Source::Path(path_source)
            }
        }
    }
}

impl From<UrlSource> for Source {
    fn from(url_source: UrlSource) -> Self {
        Source::Url(url_source)
    }
}
impl From<PathSource> for Source {
    fn from(path_source: PathSource) -> Self {
        Source::Path(path_source)
    }
}

impl Display for Source {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Source::Url(url_source) => {
                let sha256 = url_source
                    .sha256
                    .as_ref()
                    .map_or("".to_string(), |s| s.to_string());
                write!(f, "url: {}, sha256: {}", url_source.url, sha256)
            }
            Source::Path(path_source) => {
                let sha256 = path_source
                    .sha256
                    .as_ref()
                    .map_or("".to_string(), |s| s.to_string());
                write!(f, "path: {}, sha256: {}", path_source.path, sha256)
            }
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct UrlSource {
    pub url: Value<String>,
    pub sha256: Option<Value<String>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PathSource {
    pub path: Value<String>,
    pub sha256: Option<Value<String>>,
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Script {
    pub content: Vec<String>,
    pub env: IndexMap<String, String>,
    pub secrets: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NoArchKind {
    Python,
    Generic,
}

/// Python specific build configuration
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Python {
    /// For a Python noarch package to have executables it is necessary to specify the python entry points.
    /// These contain the name of the executable and the module + function that should be executed.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub entry_points: Vec<EntryPoint>,
}

impl Python {
    /// Returns true if this is the default python configuration.
    pub fn is_default(&self) -> bool {
        self.entry_points.is_empty()
    }
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Build {
    pub number: Option<Value<u64>>,
    pub script: Script,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub noarch: Option<NoArchKind>,
    #[serde(default, skip_serializing_if = "Python::is_default")]
    pub python: Python,
}

impl Build {
    pub fn new(content: Vec<String>) -> Self {
        Build {
            number: None,
            script: Script {
                content,
                ..Default::default()
            },
            ..Default::default()
        }
    }
}

/// A struct to hold the fully resolved, non-conditional requirements.
#[derive(Default)]
pub struct ResolvedRequirements {
    pub build: Vec<PackageDependency>,
    pub host: Vec<PackageDependency>,
    pub run: Vec<PackageDependency>,
    pub run_constraints: Vec<PackageDependency>,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone)]
pub enum Target {
    Default,
    Specific(String),
}

/// A type that is very specific to rattler-build /recipe.yaml side
#[derive(Serialize, Deserialize, Default, Debug)]
pub struct ConditionalRequirements {
    pub build: ConditionalList<PackageDependency>,
    pub host: ConditionalList<PackageDependency>,
    pub run: ConditionalList<PackageDependency>,
    pub run_constraints: ConditionalList<PackageDependency>,
}

impl ConditionalRequirements {
    /// Resolves the conditional requirements for a given platform.
    pub fn resolve(
        &self,
        platform: Option<&Platform>,
    ) -> PackageSpecDependencies<PackageDependency> {
        PackageSpecDependencies {
            build: self.resolve_list(&self.build, platform),
            host: self.resolve_list(&self.host, platform),
            run: self.resolve_list(&self.run, platform),
            run_constraints: self.resolve_list(&self.run_constraints, platform),
        }
    }

    pub(crate) fn resolve_list(
        &self,
        list: &ConditionalList<PackageDependency>,
        platform: Option<&Platform>,
    ) -> IndexMap<PackageName, PackageDependency> {
        list.iter()
            .flat_map(|item| Self::resolve_item(item, platform))
            .collect()
    }

    pub(crate) fn resolve_item(
        item: &Item<PackageDependency>,
        platform: Option<&Platform>,
    ) -> IndexMap<PackageName, PackageDependency> {
        match item {
            Item::Value(v) => {
                // Should we handle jinja here?
                if let Some(dep) = v.concrete() {
                    IndexMap::from([(dep.package_name(), dep.clone())])
                } else {
                    IndexMap::new()
                }
            }

            Item::Conditional(cond) => {
                if let Some(p) = platform {
                    // This is a simple string comparison
                    let dependencies = if cond.condition == *p.as_str() {
                        cond.then.clone().0.to_vec()
                    } else {
                        cond.else_value.clone().0.to_vec()
                    };

                    let mut map: IndexMap<PackageName, PackageDependency> = IndexMap::new();
                    for dep in dependencies {
                        map.insert(dep.package_name(), dep.clone());
                    }

                    map
                } else {
                    // If no platform is specified, conditional blocks are ignored.
                    IndexMap::new()
                }
            }
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct Requirements {
    pub build: Vec<SerializableMatchSpec>,
    pub host: Vec<SerializableMatchSpec>,
    pub run: Vec<SerializableMatchSpec>,
    pub run_constraints: Vec<SerializableMatchSpec>,
}

#[derive(Serialize, Deserialize)]
pub struct Test {
    pub package_contents: Option<PackageContents>,
}

#[derive(Serialize, Deserialize)]
pub struct PackageContents {
    pub include: Option<ConditionalList<String>>,
    pub files: Option<ConditionalList<String>>,
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct About {
    pub homepage: Option<Value<String>>,
    pub license: Option<Value<String>>,
    pub license_file: Option<Value<String>>,
    pub summary: Option<Value<String>>,
    pub description: Option<Value<String>>,
    pub documentation: Option<Value<String>>,
    pub repository: Option<Value<String>>,
}

#[derive(Serialize, Deserialize, Default)]
pub struct Extra {
    #[serde(rename = "recipe-maintainers")]
    pub recipe_maintainers: ConditionalList<String>,
}

// Implementation for Recipe
impl IntermediateRecipe {
    /// Converts the recipe to YAML string
    pub fn to_yaml(&self) -> Result<String, serde_yaml::Error> {
        serde_yaml::to_string(self)
    }

    /// Converts the recipe to pretty-formatted YAML string
    pub fn to_yaml_pretty(&self) -> Result<String, serde_yaml::Error> {
        // serde_yaml doesn't have a "pretty" option like serde_json,
        // but it produces readable YAML by default
        self.to_yaml()
    }

    /// Creates a recipe from YAML string
    pub fn from_yaml(yaml: &str) -> Result<IntermediateRecipe, serde_yaml::Error> {
        serde_yaml::from_str(yaml)
    }
}

impl<T: ToString + Default + Debug> Conditional<T> {
    pub fn new(condition: String, then_value: ListOrItem<T>) -> Self {
        Self {
            condition,
            then: then_value,
            else_value: ListOrItem::default(),
        }
    }

    pub fn with_else(mut self, else_value: ListOrItem<T>) -> Self {
        self.else_value = else_value;
        self
    }
}

impl<T: ToString> Value<T> {
    pub fn is_template(&self) -> bool {
        matches!(self, Value::Template(_))
    }

    pub fn is_concrete(&self) -> bool {
        matches!(self, Value::Concrete(_))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_recipe_to_yaml() {
        // Create a simple recipe
        let mut context = IndexMap::new();
        context.insert("name".to_string(), Value::Concrete("xtensor".to_string()));
        context.insert("version".to_string(), Value::Concrete("0.24.6".to_string()));

        let source = ConditionalList::from(vec![<Source as Into<Item<Source>>>::into(
            UrlSource {
                url: "https://github.com/xtensor-stack/xtensor/archive/${{ version }}.tar.gz"
                    .parse()
                    .unwrap(),
                sha256: Some(
                    "f87259b51aabafdd1183947747edfff4cff75d55375334f2e81cee6dc68ef655"
                        .parse()
                        .unwrap(),
                ),
            }
            .into(),
        )]);

        let recipe = IntermediateRecipe {
            context,
            package: Package {
                name: Value::Template("${{ name|lower }}".to_string()),
                version: Value::Template("${{ version }}".to_string()),
            },
            source,
            build: Build::default(),
            requirements: ConditionalRequirements {
                build: vec![
                    "${{ compiler('cxx') }}".parse().unwrap(),
                    "cmake".parse().unwrap(),
                    Conditional {
                        condition: "unix".to_owned(),
                        then: "make".parse().unwrap(),
                        else_value: "ninja".parse().unwrap(),
                    }
                    .into(),
                ],
                host: vec![
                    "xtl >=0.7,<0.8".parse().unwrap(),
                    "${{ context.name }}".parse().unwrap(),
                ],
                run: vec!["xtl >=0.7,<0.8".parse().unwrap()],
                run_constraints: vec!["xsimd >=8.0.3,<10".parse().unwrap()],
            },
            about: Some(About {
                homepage: Some(Value::Concrete(
                    "https://github.com/xtensor-stack/xtensor".to_string(),
                )),
                license: Some("BSD-3-Clause".parse().unwrap()),
                license_file: Some("LICENSE".parse().unwrap()),
                summary: Some("The C++ tensor algebra library".parse().unwrap()),
                description: Some(
                    "Multi dimensional arrays with broadcasting and lazy computing"
                        .parse()
                        .unwrap(),
                ),
                documentation: Some("https://xtensor.readthedocs.io".parse().unwrap()),
                repository: Some("https://github.com/xtensor-stack/xtensor".parse().unwrap()),
            }),
            extra: Some(Extra {
                recipe_maintainers: vec!["some-maintainer".parse().unwrap()],
            }),
            ..Default::default()
        };

        insta::assert_yaml_snapshot!(recipe)
    }
}
