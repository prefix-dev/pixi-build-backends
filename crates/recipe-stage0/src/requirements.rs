use std::collections::HashSet;

use indexmap::IndexMap;
use rattler_build::NormalizedKey;
use rattler_conda_types::PackageName;

use crate::matchspec::PackageDependency;

/// A package spec dependency represent dependencies for a specific target.
#[derive(Debug, Clone)]
pub struct PackageSpecDependencies<T> {
    pub build: IndexMap<PackageName, T>,
    pub host: IndexMap<PackageName, T>,
    pub run: IndexMap<PackageName, T>,
    pub run_constraints: IndexMap<PackageName, T>,
}

impl<T> Default for PackageSpecDependencies<T> {
    fn default() -> Self {
        PackageSpecDependencies {
            build: IndexMap::new(),
            host: IndexMap::new(),
            run: IndexMap::new(),
            run_constraints: IndexMap::new(),
        }
    }
}

impl PackageSpecDependencies<PackageDependency> {
    /// Return the used variants of the package spec dependencies.
    pub fn used_variants(&self) -> HashSet<NormalizedKey> {
        let used_variants = self
            .build
            .iter()
            .chain(self.host.iter())
            .chain(self.run.iter())
            .filter(|(_, spec)| spec.can_be_used_as_variant())
            .map(|(name, _)| name.clone().as_normalized().into())
            .collect();

        used_variants
    }

    pub fn contains(&self, name: &PackageName) -> bool {
        self.build.contains_key(name)
            || self.host.contains_key(name)
            || self.run.contains_key(name)
            || self.run_constraints.contains_key(name)
    }
}

/// Represents a platform, selector.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub enum Selector {
    Unix,
    Linux,
    Win,
    MacOs,
    Platform(String),
}
