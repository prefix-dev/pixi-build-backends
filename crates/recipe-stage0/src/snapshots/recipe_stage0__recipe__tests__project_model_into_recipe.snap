---
source: crates/recipe-stage0/src/recipe.rs
expression: ir
---
context: {}
package:
  name: minimal-package
  version: 1.0.0
source:
  - path: /path/to/manifest
    sha256: ~
build:
  number: ~
  script: []
requirements:
  build: []
  host:
    - if: linux
      then: boltons
      else: []
  run:
    - python
  run_constraints: []
tests: []
about: ~
extra: ~
